src/clojure/nrepl/socket.clj

```diff
@@ -0,0 +1,222 @@
+(ns nrepl.socket
+  "Compatibility layer for java.io vs java.nio sockets to allow an
+  incremental transition to nio, since the JDK's filesystem sockets
+  don't support the java.io socket interface, and we can't use the
+  compatibility layer for bidirectional read and write:
+  https://bugs.openjdk.java.net/browse/JDK-4509080."
+  (:require
+   [clojure.java.io :as io]
+   [nrepl.misc :refer [log]])
+  (:import
+   (clojure.lang Reflector)
+   (java.io BufferedInputStream BufferedOutputStream File)
+   (java.net InetSocketAddress ServerSocket Socket StandardProtocolFamily URI)
+   (java.nio ByteBuffer)
+   (java.nio.file Path)
+   (java.nio.channels Channels ClosedChannelException
+                      ServerSocketChannel SocketChannel)))
+
+(def orig-warn-on-reflection *warn-on-reflection*)
+
+(defmacro find-class [full-path]
+  `(try
+     (import ~full-path)
+     (catch ClassNotFoundException ex#
+       nil)))
+
+;;; InetSockets (TCP)
+
+(defn inet-socket [bind port]
+  (let [port (or port 0)
+        addr (fn [^String bind ^Integer port] (InetSocketAddress. bind port))
```

Comment:
The arg is int not Integer https://docs.oracle.com/javase/7/docs/api/java/net/InetSocketAddress.html

In any case, I wouldn't use hints here but casting, it's what's recommended for primitives:

![image](https://user-images.githubusercontent.com/1162994/125428895-3d3f106b-c931-412c-94ee-0a87f9e60557.png)

Reply:
Thanks, fix pending.